### ISSUE TITLE
Image data from training + inference load functions don't match

### ISSUE THREAD

**Issue by addisonklinke** [JDAI-CV/fast-reid#298]
## Instructions To Reproduce the 🐛 Bug:

1. what changes you made (`git diff`) or what code you wrote

Starting from the latest code at caf11b5c

- Remove `PRETRAIN_PATH` from Market1501 BoT config so PyTorch will download to my user's home
- Update `CommDataset.__getitem__` to save the numpy image for a test set example to disk
- Use only the first 64 images for training (to get to the evaluation loop faster)

```
diff --git a/configs/Base-bagtricks.yml b/configs/Base-bagtricks.yml
index 43d2af2..8d8190d 100644
--- a/configs/Base-bagtricks.yml
+++ b/configs/Base-bagtricks.yml
@@ -9,7 +9,6 @@ MODEL:
     FEAT_DIM: 2048
     WITH_IBN: False
     PRETRAIN: True
-    PRETRAIN_PATH: "/export/home/lxy/.cache/torch/checkpoints/resnet50-19c8e357.pth"
 
   HEADS:
     NAME: "EmbeddingHead"
diff --git a/fastreid/data/common.py b/fastreid/data/common.py
index 82feb3b..8f5d8bf 100644
--- a/fastreid/data/common.py
+++ b/fastreid/data/common.py
@@ -4,6 +4,7 @@
 @contact: sherlockliao01@gmail.com
 """
 
+import numpy as np
 from torch.utils.data import Dataset
 
 from .data_utils import read_image
@@ -39,6 +40,10 @@ class CommDataset(Dataset):
         if self.relabel:
             pid = self.pid_dict[pid]
             camid = self.cam_dict[camid]
+        example_img = '0000_c1s1_000151_01.jpg'
+        if img_path.endswith(example_img):
+            print('Saving numpy image to disk!')
+            np.save(f'/tmp/{example_img}.npy', img)
         return {
             "images": img,
             "targets": pid,
diff --git a/fastreid/data/datasets/market1501.py b/fastreid/data/datasets/market1501.py
index d1968af..6a49553 100644
--- a/fastreid/data/datasets/market1501.py
+++ b/fastreid/data/datasets/market1501.py
@@ -68,7 +68,7 @@ class Market1501(ImageDataset):
         if self.market1501_500k:
             gallery += self.process_dir(self.extra_gallery_dir, is_train=False)
 
-        super(Market1501, self).__init__(train, query, gallery, **kwargs)
+        super(Market1501, self).__init__(train[:64], query, gallery, **kwargs)
 
     def process_dir(self, dir_path, is_train=True):
         img_paths = glob.glob(osp.join(dir_path, '*.jpg'))
```
2. what exact command you run:

First, run the training script with modifications to save the numpy cache

```
python tools/train_net.py --config-file ./configs/Market1501/bagtricks_R50.yml
```

Then, run a simple script that uses the `preprocess` function in `tools/deploy/onnx_inference` to load the same image and compare it to the numpy cache from training

```
import numpy as np
from tools.deploy.onnx_inference import preprocess

# Load the same image with two different methods
# Numpy cache from training vs. inference preprocess
# 256 x 128 dimensions come from cfg.INPUT.SIZE_TRAIN
# Inference is a 4D mini-batch, so squeeze it down to 3D
example_img = '0000_c1s1_000151_01.jpg'
img_path = './datasets/Market-1501-v15.09.15/bounding_box_test/' + example_img
training = np.load(f'/tmp/{example_img}.npy')
inference = preprocess(img_path, 256, 128).squeeze()

# Check whether the pixel data is a perfect match
print('Image tensor shapes')
print(f'Inference: {inference.shape}')
print(f'Training: {training.shape}')
matching_pixels = (training == inference).sum()
total_pixels = training.size
print(f'{matching_pixels / total_pixels * 100:.2f}% of pixels match')
```

3. what you observed (including __full logs__):

The comparison output shows that the inference vs. training loading methods do not produce the same image

```
Image tensor shapes
Inference: (3, 256, 128)
Training: (3, 256, 128)
42.85% of pixels match
```

Training log included below

```
Command Line Args: Namespace(config_file='./configs/Market1501/bagtricks_R50.yml', dist_url='tcp://127.0.0.1:50152', eval_only=False, finetune=False, machine_rank=0, num_gpus=1, num_machines=1, opts=[], resume=False)
[10/07 11:55:01 fastreid]: Rank of current process: 0. World size: 1
[10/07 11:55:01 fastreid]: Environment info:
    ... already included below ...
[10/07 11:55:01 fastreid]: Command line arguments: Namespace(config_file='./configs/Market1501/bagtricks_R50.yml', dist_url='tcp://127.0.0.1:50152', eval_only=False, finetune=False, machine_rank=0, num_gpus=1, num_machines=1, opts=[], resume=False)
[10/07 11:55:01 fastreid]: Contents of args.config_file=./configs/Market1501/bagtricks_R50.yml:
    ... omitted to save space ...
[10/07 11:55:01 fastreid]: Full config saved to /storage/projects/alpr/modules/fast-reid/logs/market1501/bagtricks_R50/config.yaml
[10/07 11:55:01 fastreid.utils.env]: Using a generated random seed 1753435
[10/07 11:55:01 fastreid.engine.defaults]: Prepare training set
[10/07 11:55:01 fastreid.data.datasets.bases]: => Loaded Market1501 in csv format: 
| subset   | # ids   | # images   | # cameras   |
|:---------|:--------|:-----------|:------------|
| train    | 60      | 64         | 6           |
[10/07 11:55:01 fastreid.engine.defaults]: Auto-scaling the config to num_classes=60, max_Iter=120, wamrup_Iter=10, freeze_Iter=0, delay_Iter=0, step_Iter=[40, 90], ckpt_Iter=400, eval_Iter=200.
[10/07 11:55:02 fastreid.modeling.backbones.resnet]: Loading pretrained model from /home/addison/.cache/torch/checkpoints/resnet50-19c8e357.pth
[10/07 11:55:02 fastreid.modeling.backbones.resnet]: The checkpoint contains parameters not used by the model:
  fc.{weight, bias}
[10/07 11:55:03 fastreid.engine.defaults]: Model:
    ... omitted to save space ...
[10/07 11:55:03 fastreid.utils.checkpoint]: No checkpoint found. Training model from scratch
[10/07 11:55:03 fastreid.engine.train_loop]: Starting training from iteration 0
[10/07 11:56:30 fastreid.utils.checkpoint]: Saving checkpoint to logs/market1501/bagtricks_R50/model_final.pth
[10/07 11:56:30 fastreid.engine.defaults]: Prepare testing set
[10/07 11:56:31 fastreid.data.datasets.bases]: => Loaded Market1501 in csv format: 
| subset   | # ids   | # images   | # cameras   |
|:---------|:--------|:-----------|:------------|
| query    | 750     | 3368       | 6           |
| gallery  | 751     | 15913      | 6           |
[10/07 11:56:31 fastreid.evaluation.evaluator]: Start inference on 19281 images
[10/07 11:56:38 fastreid.evaluation.evaluator]: Inference done 11/151. 0.0077 s / batch. ETA=0:01:03
[10/07 11:57:09 fastreid.evaluation.evaluator]: Inference done 65/151. 0.0092 s / batch. ETA=0:00:47
Saving numpy image to disk!
[10/07 11:57:39 fastreid.evaluation.evaluator]: Inference done 119/151. 0.0093 s / batch. ETA=0:00:17
[10/07 11:57:59 fastreid.evaluation.evaluator]: Total inference time: 0:01:23.352315 (0.570906 s / batch per device)
[10/07 11:57:59 fastreid.evaluation.evaluator]: Total inference pure compute time: 0:00:03 (0.021654 s / batch per device)
[10/07 11:58:07 fastreid.evaluation.testing]: Evaluation results in csv format: 
| Datasets   | Rank-1   | Rank-5   | Rank-10   | mAP   | mINP   |
|:-----------|:---------|:---------|:----------|:------|:-------|
| Market1501 | 7.21%    | 15.02%   | 19.89%    | 2.45% | 0.30%  |
[10/07 11:58:07 fastreid.utils.events]:  eta: 0:00:00  iter: 119  total_loss: 0.8271  loss_cls: 0.8271  loss_triplet: 0  time: 0.6951  data_time: 0.0007  lr: 3.50e-06  max_mem: 4464M
[10/07 11:58:07 fastreid.engine.hooks]: Overall training speed: 117 iterations in 0:01:22 (0.7010 s / it)
[10/07 11:58:07 fastreid.engine.hooks]: Total training time: 0:02:58 (0:01:36 on hooks)
```
4. please simplify the steps as much as possible so they do not require additional resources to
	 run, such as a private dataset.

Done, this is run using the built-in Market1501

## Expected behavior:

An image should have identical pixel content whether it is loaded with the training or inference code. Otherwise the model will not produce the same embeddings since the input values are different. In order to successfully write an inference program, we need a function that loads and/or pre-processes the image data following the exact protocol used in the training/testing loops. Why does the current code not accomplish this?

## Environment:

```
sys.platform            linux
Python                  3.6.11 | packaged by conda-forge | (default, Aug  5 2020, 20:09:42) [GCC 7.5.0]
numpy                   1.19.1
detectron2              failed to import
detectron2._C           failed to import. detectron2 is not built correctly
Compiler                c++ (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
CUDA compiler           Cuda compilation tools, release 10.0, V10.0.130
DETECTRON2_ENV_MODULE   <not set>
PyTorch                 1.6.0 @/home/addison/miniconda3/envs/fastreid/lib/python3.6/site-packages/torch
PyTorch debug build     False
GPU available           True
GPU 0                   GeForce GTX 1060 with Max-Q Design (arch=6.1)
CUDA_HOME               /usr/local/cuda
Pillow                  7.2.0
torchvision             0.7.0 @/home/addison/miniconda3/envs/fastreid/lib/python3.6/site-packages/torchvision
torchvision arch flags  3.5, 5.0, 6.0, 7.0, 7.5
cv2                     4.4.0
----------------------  -------------------------------------------------------------------------------------
PyTorch built with:
  - GCC 7.3
  - C++ Version: 201402
  - Intel(R) Math Kernel Library Version 2020.0.2 Product Build 20200624 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v1.5.0 (Git Hash e2ac1fac44c5078ca927cb9b90e1b3066a0b2ed0)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 10.2
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_37,code=compute_37
  - CuDNN 7.6.5
  - Magma 2.5.2
  - Build settings: BLAS=MKL, BUILD_TYPE=Release, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DUSE_VULKAN_WRAPPER -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, USE_CUDA=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, USE_STATIC_DISPATCH=OFF,
```

**Comment by L1aoXingyu** [comment]
The data augmentations in the training phase and testing phase are different, so you cannot get the same result.

**Comment by addisonklinke** [comment]
@L1aoXingyu I realize the training images are different due to the augmentations applied. However, I purposely cached a test set image to sidestep this. The test images are not augmented (other than resizing the pixel dimensions), so they should match what is provided during inference. In hindsight, I should have worded my issue title and code differently to make this clear.

After examining this in more detail, I believe the difference is train/test images are loaded using `PIL` while inference images use `cv2`. This [article](https://towardsdatascience.com/image-read-and-resize-with-opencv-tensorflow-and-pil-3e0f29b992be) explains that the two image libraries have slightly different decompression algorithms for JPEG data:

> Well, this difference is arising from the fact that OpenCV, by default, uses integer accurate decompression of the JPEG image ... PIL, by default, doesn’t provide any way to use integer accurate decompression

Therefore, I would recommend using a consistent library for train, test, and inference. My preference would be OpenCV since it also supports C++ which would make comparisons to production code easier

**Comment by L1aoXingyu** [comment]
Thanks for your suggestions so much. This is exactly the reason why pixel errors between training and testing.
The use of `torchvision.transforms.Resize` in training is compatible with other built-in data augmentation. But in testing, we need to export to c++, so the only choice is opencv.
I am considering using opencv for both training and inference.

**Comment by addisonklinke** [comment]
Sounds good - I think we have an explanation for the current behavior, so I'll close this issue now. If you end up using OpenCV for both, please comment back once the change is made
